### ISSUE TITLE
Antialiasing Support (GraphicsDevice.PreferMultiSampling)

### ISSUE THREAD

**Issue by ghost** [MonoGame/MonoGame#358]
All platforms need to have antialiasing (re-?)enabled.  Pointers [here](http://www.gandogames.com/2010/07/tutorial-using-anti-aliasing-msaa-in-the-iphone/) and [here](https://devforums.apple.com/thread/45850).

**Comment by aienabled** [comment]
I'm using DirectX11 and changing `GraphicsDevice.PresentationParameters.MultiSampleCount` during runtime has no effect (`Apply()` is called).
However, if I set this in MonoGame source code (method `GraphicsDeviceManager.Initialize()`) before creating the graphics device, it works fine.
So the issue is only with applying the multi-sampling after initialization.

**Comment by GuilhermeAmorim93** [comment]
I'm using DirectX11 too,
But changing it during Game.Initialize() with `GraphicsDevice.PresentationParameters.MultiSampleCount = 8; GraphicsDeviceManager.Apply();`
causes nothing to be rendered. 

Any value other than 1 causes this to happen.

**Comment by Noemata** [comment]
Latest build still has this issue.

**Comment by SeriousMartin** [comment]
This problem still persists with the current version (https://github.com/MonoGame/MonoGame/commit/dd2e6cb79ecf6cb553582c402635d13faf9e9596 with #5477 merged). Any idea what could cause this?

**Comment by zigzag312** [comment]
@SeriousMartin On what platform and are you using render targets? #5477 fixes only one small part of AA issues.

**Comment by Jjagg** [comment]
@SeriousMartin You mean for a DirectX project? If so can you check the value of `MaxMultiSampleCount` in `GraphicsDevice.GraphicsCapabilities` for your `GraphicsDevice` using a debugger. Are you setting `GraphicsDeviceManager.PreferMultiSampling = true` (either in your constructor or by calling `ApplyChanges` afterwards) or using the `OnPreparingDeviceSettings` event? Note that the method mentioned above that directly changes `GraphicsDevice.PresentationParameters.MultiSampleCount` does not and should not work.
The three correct ways for modifying multisample count - and `PresentationParameters` in general - are the following:

- Setting `GraphicsDeviceManager.PreferMultiSampling` will set multisample count to the maximum supported by the GPU
- Subscribing to the `GraphicsDeviceManager.PreparingDeviceSettings` event and modifying the `PresentationParameters.MultiSampleCount` on the event args will set the ms count to the minimum of that value and what's supported by the device.
- To modify `PresentationParameters` of a `GraphicsDevice`, first prepare a `PresentationParameters` instance with whatever `MultiSampleCount` you want, then call `GraphicsDevice.Reset(PresentationParameters)` with that instance. This will have the same effect as the previous option.

All three of these should work for DirectX (there's some issues with the last two options for DesktopGL because the `GameWindow` needs to be recreated), if they don't something is borken™.

**Comment by Jjagg** [comment]
> On what platform and are you using render targets

This issue is about multisampling on the back buffer, there's separate issues for `RenderTarget`s.

**Comment by SeriousMartin** [comment]
Yes, DirectX. What I'm trying to do is to activate/deactivate multi sampling during runtime.

I have create a new MonoGame project rendering only a sprite onto the screen. 
Then I added the following code to the Update method:

`            if (Keyboard.GetState().IsKeyDown(Keys.M) && !graphics.PreferMultiSampling)
            {
                graphics.PreferMultiSampling = true;
                graphics.ApplyChanges();
            }
            if (Keyboard.GetState().IsKeyDown(Keys.N) && graphics.PreferMultiSampling)
            {
                graphics.PreferMultiSampling = false;
                graphics.ApplyChanges();
            }`

The game starts with multi sampling deactivated. When I press M once the sprite is no longer visible. When I press N it's visible again.

If I start the game with multi sampling activated it's vice versa.

I'd expect this to work. Did I get something wrong?

**Comment by Jjagg** [comment]
> Did I get something wrong?

No, that should work. This is still bugged then...

**Comment by zigzag312** [comment]
@SeriousMartin Would you mind uploading that sample somewhere?

**Comment by SeriousMartin** [comment]
@zigzag312 https://www.dropbox.com/s/yzlbem61wtp09rn/Game1.zip?dl=0
It's a visual studio 2015 solution

**Comment by zigzag312** [comment]
@SeriousMartin Thanks for the sample. I have found source of your problem. 
This bug only happens when changing multisampling mid-game. 

In GraphicsDevice.DirectX.cs -> CreateSizeDependentResources() `_swapChain `is only resized if it already exists, but depth buffer is always created form scratch. This results in mismatch between swap chain and depth buffer, because `SampleDescription `is not correctly updated for swap chain. End result is that nothing is visible. 

I'll create PR fix for this.

**Comment by CartBlanche** [comment]
@tomspilman @Jure-BB etc Is this still an issue in Direct X or only OpenGL now? I noticed this has been pushed back to 3.9.x now?
